### PR TITLE
[ci, build] Check in the Cargo.lock + add a workspace-hack package

### PR DESCRIPTION
**Pull request by huitseeker**
The workspace-hack is a now-standard technique to pull less dependencies in a build. See the doc for rationale:
https://docs.rs/cargo-hakari/0.9.6/cargo_hakari/about/index.html#what-are-workspace-hack-crates

What's interesting is a gajillion years ago, [we had the intuition such a package could be used to pull in all the dependencies](https://github.com/MystenLabs/sui-operations/pull/1#discussion_r854505647) involved in a build in one go, which would create a stable layer in a Docker build.

Today, @bmwill has verified this intuition:
https://github.com/MystenLabs/sui/pull/2246

This PR creates a workspace package in Narwhal so we can use the same techniques in our docker build, speeding up the re-builds there through proper layering / caching.

See here for a successful CI run
https://github.com/huitseeker/narwhal/runs/6603591773?check_suite_focus=true

### CODE REVIEW

**Review by akichidis — APPROVED** [review]
Thanks @huitseeker for this improvement!

If understand correctly, we'll mostly benefit from this on our local rebuilds, rather than the CI ones, right?

File: node/Cargo.toml [hunk 1]
```diff
@@ -32,6 +32,8 @@ executor = { path = "../executor" }
 primary = { path = "../primary" }
 types = { path = "../types" }
 worker = { path = "../worker" }
+mirai-annotations = "1.12.0"
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply node/Cargo.toml]: good spot! This is for a different experiment I was running that has nothing to do with this PR, I'll clean that up.

**Comment by huitseeker** [comment]
> If understand correctly, we'll mostly benefit from this on our local rebuilds, rather than the CI ones, right?

We benefit equally on both. The CI components of this PR are there just to check the machinery that maintains the structure of the dependencies in the project isn't broken by accident.

**Comment by akichidis** [comment]
> > If understand correctly, we'll mostly benefit from this on our local rebuilds, rather than the CI ones, right?
> 
> We benefit equally on both. The CI components of this PR are there just to check the machinery that maintains the structure of the dependencies in the project isn't broken by accident.

Oh yes absolutely, but is not like we should expect to speed up the CI builds, correct? Or am I missing something?

**Comment by huitseeker** [comment]
@akichidis 
> Oh yes absolutely, but is not like we should expect to speed up the CI builds, correct? Or am I missing something?

Yes, we should. See the rationale in the first link of the PR description, the explanation there is really better than what I could pen here.

**Pull request merged**